lib/vtls/schannel.c

```diff
@@ -708,22 +727,235 @@ schannel_acquire_credential_handle(struct Curl_easy *data,
-                                       &backend->cred->time_stamp);
+  /* Windows 10, 1809 (a.k.a. Windows 10 build 17763) */
+  if(curlx_verify_windows_version(10, 0, 17763, PLATFORM_WINNT,
+      VERSION_GREATER_THAN_EQUAL)) {
+
+    int disable_aes_gcm_sha384 = 0;
+    int disable_aes_gcm_sha256 = 0;
+    int disable_chacha_poly = 1;
+
+    SCH_CREDENTIALS credentials = { 0 };
+    TLS_PARAMETERS tls_parameters = { 0 };
+    CRYPTO_SETTINGS crypto_settings[4] = { 0 };
+    UNICODE_STRING blocked_ccm_modes[1] = { 0 };
+    UNICODE_STRING blocked_gcm_modes[1] = { 0 };
+
+    /*
+     RTM Windows 2022 / Windows 11 does not support ChaCha20-Poly1305
+     or AES_CCM. So disable both those options explictly (TLS 1.3
+     includes those as 2 built-in Cipher-Suites).
+
+     Add conditional curlx_verify_windows_version() calls when
+     support has been added.
+    */
+
+    int crypto_settings_idx = 0;
 
+
+    /*
+      Disallow AES_CCM algorithm, since there's no support for it in SChannel.
+    */
+    blocked_ccm_modes[0] = (UNICODE_STRING) {
+        sizeof(BCRYPT_CHAIN_MODE_CCM),
+        sizeof(BCRYPT_CHAIN_MODE_CCM),
+        BCRYPT_CHAIN_MODE_CCM };
+
+    crypto_settings[crypto_settings_idx].eAlgorithmUsage =
+                                             TlsParametersCngAlgUsageCipher;
+    crypto_settings[crypto_settings_idx].rgstrChainingModes =
+                                             blocked_ccm_modes;
+    crypto_settings[crypto_settings_idx].cChainingModes =
+                                             ARRAYSIZE(blocked_ccm_modes);
+    crypto_settings[crypto_settings_idx].strCngAlgId = (UNICODE_STRING) {
+        sizeof(BCRYPT_AES_ALGORITHM),
+        sizeof(BCRYPT_AES_ALGORITHM),
+        BCRYPT_AES_ALGORITHM };
+    crypto_settings_idx++;
+
+
+    /* Only 3 ciphers for TLS 1.3 remain:
+    * TLS_AES_256_GCM_SHA384
+    * TLS_AES_128_GCM_SHA256
+    * TLS_CHACHA20_POLY1305_SHA256
+    */
+    char *ciphers13 = SSL_CONN_CONFIG(cipher_list13);
+    if(ciphers13) {
+      int remaining_ciphers = 3;
+
+      /* detect which remaining ciphers to enable
+      and then disable everything else.
+      */
+
+      char *startCur = ciphers13;
+      int algCount = 0;
+      char tmp[LONGEST_ALG_ID] = { 0 };
+      char *nameEnd;
+
+      disable_aes_gcm_sha384 = 1;
+      disable_aes_gcm_sha256 = 1;
+
+      while(startCur && (0 != *startCur) && (algCount < remaining_ciphers)) {
+
+        nameEnd = strchr(startCur, ':');
+        size_t n = nameEnd ? \
+            min((size_t)(nameEnd - startCur), LONGEST_ALG_ID - 1) : \
+            min(strlen(startCur), LONGEST_ALG_ID - 1);
+        strncpy(tmp, startCur, n);
+        tmp[n] = 0;
+
+        if(strcmp("TLS_AES_256_GCM_SHA384", tmp) == 0)
+          disable_aes_gcm_sha384 = 0;
+        else if(strcmp("TLS_AES_128_GCM_SHA256", tmp) == 0)
+          disable_aes_gcm_sha256 = 0;
+        else if(strcmp("TLS_CHACHA20_POLY1305_SHA256", tmp) == 0)
+          disable_chacha_poly = 0;
```

Comment:
If this cipher suite isn't supported, isn't this then just going to fail?

Reply:
Only if the user left it disabled at the system level. They can, however, [enable it](https://docs.microsoft.com/en-us/powershell/module/tls/enable-tlsciphersuite?view=windowsserver2022-ps). Also, the other ciphers (`TLS_AES_128_GCM_SHA256` for example) can be disabled at the system level and also fail.

So, yes, a customer can shoot themselves in the foot several ways. The default behavior I chose was what was enabled by default in Windows (ChCha-Poly disabled, and the AES_GCM ciphers enabled).